pep425.nix

```diff
@@ -73,9 +74,9 @@ let
         if isLinux
         then
           (
-            x: x.platform == "manylinux1_${stdenv.platform.kernelArch}"
-              || x.platform == "manylinux2010_${stdenv.platform.kernelArch}"
-              || x.platform == "manylinux2014_${stdenv.platform.kernelArch}"
+            x: x.platform == "manylinux1_${targetArch}"
+              || x.platform == "manylinux2010_${targetArch}"
+              || x.platform == "manylinux2014_${targetArch}"
```

Comment:
We can just use `stdenv.hostPlatform.parsed.cpu.arch` instead, no need to introduce this into `poetryLib` then, and that's compatible with both stable and unstable.

Reply:
I don't think this works: for example, x86_64 has `cpu.arch` set to [`x86-64`](https://github.com/NixOS/nixpkgs/blob/8baac2af752450b20cb8b37520059ecd86490b8b/lib/systems/parse.nix#L89), whereas we specifically want `x86_64`. Now that I'm looking at it, I'm pretty sure `linuxArch` doesn't work either - for example, it uses [`arm`](https://github.com/NixOS/nixpkgs/blob/8baac2af752450b20cb8b37520059ecd86490b8b/lib/systems/default.nix#L90) for aarch32, but PyPI uses `armv7l`. I guess the most correct solution would be to introduce a custom mapping from `stdenv.hostPlatform.parsed.cpu` to PyPI tags...

Reply:
Ah I think we can use `stdenv.hostPlatform.parsed.cpu.name` then

Reply:
Not quite - `powerpc64` is `ppc64` on PyPI :(

Reply:
Where do you see that?

Reply:
Python ABI tags use the exact architecture specifiers `uname` uses, so Linux's `ARCH`. For manylinux2014, they're listed here: https://www.python.org/dev/peps/pep-0599/#the-manylinux2014-policy

Reply:
More specifically, we're looking at

```
x86_64
i686
aarch64
armv7l
ppc64
ppc64le
s390x
```

Reply:
Nixpkgs doesn't run on s390x AFAIK, so we only need to check for 6 platforms.

Reply:
(also, manylinux1 and manylinux2010 are restricted to x86_64 and i686 only, but we can still check for every architecture, as manylinux{1,2010} wheels for other architectures can't be uploaded to PyPI)

Reply:
I think using `parsed.cpu.name` should be good enough for now. It's correct for the most important ones, and notably also for more than the previous `stdenv.platform.kernelArch`. On 20.09:
```
❯ ne -A stdenv.platform.kernelArch --argstr system x86_64-linux
"x86_64"
❯ ne -A stdenv.platform.kernelArch --argstr system i686-linux
"i386"
❯ ne -A stdenv.platform.kernelArch --argstr system aarch64-linux
"arm64"
❯ ne -A stdenv.platform.kernelArch --argstr system armv7l-linux
"arm"
```

But for `stdenv.hostPlatform.parsed.cpu.name`:
```
❯ ne -A stdenv.hostPlatform.parsed.cpu.name --argstr system x86_64-linux
"x86_64"
❯ ne -A stdenv.hostPlatform.parsed.cpu.name --argstr system i686-linux
"i686"
❯ ne -A stdenv.hostPlatform.parsed.cpu.name --argstr system aarch64-linux
"aarch64"
❯ ne -A stdenv.hostPlatform.parsed.cpu.name --argstr system armv7l-linux
"armv7l"
```

Reply:
This still gets us `powerpc` and `powerpc64` though. I'd prefer to add an explicit lookup table, especially since it's going to be very small. Is there something I'm missing that makes it a bad idea?

Reply:
I think these values should ideally come from nixpkgs, to keep platform handling out of poetry2nix. From the PEP's, it looks like these values come from CentOS (for PEP425), or from `platform.machine() -> uname -m` (for PEP508). Nixpkgs notably already has a partial `uname` mapping. These all don't seem python specific.

How about we just use `parsed.cpu.name` for now to keep it simple, but you can make a nixpkgs PR for adding such mappings (where you can also ping the people who know more about platform stuff), and once it's been in nixpkgs for a while we can switch poetry2nix to use that.

Reply:
> I think these values should ideally come from nixpkgs, to keep platform handling out of poetry2nix.

Agreed conceptually, but I'd rather have a mapping in place now and then switch to nixpkgs' one when it's good enough. Otherwise we're basically knowingly breaking PowerPC on all stable NixOS versions and then however long it takes for nixpkgs to get a proper mapping.

> From the PEP's, it looks like these values come from CentOS (for PEP425), or from platform.machine() -> uname -m (for PEP508)

The CentOS values are also just `uname -m` from CentOS machines, so this shouldn't be a problem.

> Nixpkgs notably already has a partial uname mapping. These all don't seem python specific.

I was actually looking for a `uname -m` mapping in nixpkgs and couldn't find one. Do you have a link to the one you mentioned? (I've seen [this](https://github.com/NixOS/nixpkgs/blob/8baac2af752450b20cb8b37520059ecd86490b8b/lib/systems/default.nix#L73), but that's not helpful unfortunately).

> How about we just use parsed.cpu.name for now to keep it simple, but you can make a nixpkgs PR for adding such mappings (where you can also ping the people who know more about platform stuff), and once it's been in nixpkgs for a while we can switch poetry2nix to use that.

I'm up to do that (and discuss it on the nixpkgs side), but again, I'd rather have a temporary mapping in place in poetry2nix for now, or at the very least, some sort of a comment that explicitly mentions the PowerPC issue.

Reply:
Regarding the `uname` mapping, I said "partial" because it notably doesn't include all the `uname` flags, such as `-m` :)

Considering that I don't believe anybody has even tested poetry2nix with powerpc64, I think it wouldn't be a problem to just not support it for now until we have the correct mapping in nixpkgs. So I'd just go for an assertion that checks the cpu name to be one of the supported ones.

But if you'd much rather have a mapping directly in poetry2nix, that's fine too I guess. Feel free to go for it.

Reply:
All right, I'll try to write up a patch in a few hours. Thanks!

Reply:
...actually took me a lot less than a few hours.